pkg/common/utils/utils.go

```diff
@@ -80,18 +83,44 @@ func QueryAllVolumeUtil(ctx context.Context, m cnsvolume.Manager, queryFilter cn
 	querySelection *cnstypes.CnsQuerySelection, useQueryVolumeAsync bool) (*cnstypes.CnsQueryResult, error) {
 	log := logger.GetLogger(ctx)
 	var queryAsyncNotSupported bool
-	var queryResult *cnstypes.CnsQueryResult
+	var allQueryResults []*cnstypes.CnsQueryResult
+	queryResult := &cnstypes.CnsQueryResult{
+		Volumes: make([]cnstypes.CnsVolume, 0),
+	}
 	var err error
 	if useQueryVolumeAsync {
-		// AsyncQueryVolume feature switch is disabled.
-		queryResult, err = m.QueryVolumeAsync(ctx, queryFilter, querySelection)
-		if err != nil {
-			if err.Error() == cnsvsphere.ErrNotSupported.Error() {
-				log.Warn("QueryVolumeAsync is not supported. Invoking QueryAllVolume API")
-				queryAsyncNotSupported = true
-			} else { // Return for any other failures.
-				return nil, logger.LogNewErrorCodef(log, codes.Internal,
-					"queryVolumeAsync failed for queryFilter: %v. Err=%+v", queryFilter, err.Error())
+		// AsyncQueryVolume feature switch is enabled.
+		cursor := &cnstypes.CnsCursor{
+			Offset: 0,
+			Limit:  queryVolumeLimit,
+		}
+		queryFilter.Cursor = cursor
+		// Using pagination technique to fetch the entire QueryResults
+		for {
+			queryRes, err := m.QueryVolumeAsync(ctx, queryFilter, querySelection)
+			if err != nil {
+				if err.Error() == cnsvsphere.ErrNotSupported.Error() {
+					log.Warn("QueryVolumeAsync is not supported. Invoking QueryAllVolume API")
+					queryAsyncNotSupported = true
+					break
+				} else { // Return for any other failures.
+					return nil, logger.LogNewErrorCodef(log, codes.Internal,
+						"queryVolumeAsync failed for queryFilter: %v. Err=%+v", queryFilter, err.Error())
+				}
+			}
+			allQueryResults = append(allQueryResults, queryRes)
+			log.Infof("%v more volumes to be queried", queryRes.Cursor.TotalRecords-queryRes.Cursor.Offset)
```

Comment:
This logger should be moved at line 117, so that it only gets printed when there are more volumes to be queried.

Reply:
Makes sense